tempesta_fw/http.c

```diff
@@ -796,122 +1501,272 @@ tfw_http_adjust_resp(TfwHttpResp *resp, TfwHttpReq *req)
 				     TFW_HTTP_HDR_SERVER, 0);
 }
 
+/*
+ * Forward responses in @ret_queue to the client in correct order.
+ *
+ * In case of error the client connection must be closed immediately.
+ * Otherwise, the correct order of responses will be broken. Unsent
+ * responses are taken care of by the caller.
+ */
+static void
+__tfw_http_resp_fwd(TfwCliConn *cli_conn, struct list_head *ret_queue)
+{
+	TfwHttpReq *req, *tmp;
+
+	list_for_each_entry_safe(req, tmp, ret_queue, msg.seq_list) {
+		BUG_ON(!req->resp);
+		if (tfw_cli_conn_send(cli_conn, (TfwMsg *)req->resp)) {
+			ss_close_sync(cli_conn->sk, true);
+			return;
+		}
+		__tfw_http_resp_pair_free(req);
```

Comment:
I am not sure I understand. The response (and the request) is not needed after it's sent out. If for some reason a special flag is set that tells that the SKB data is needed, then the SKBs are copied. In that case the response and it's SKB data can be freed. If SKB data is not needed (the usual case), then the full chain of SKBs is simply detached from the response. When the response is freed there's no SKBs to free.